services/m3aggregator/config/runtime.go

```diff
@@ -0,0 +1,189 @@
+func (c RuntimeOptionsConfiguration) NewRuntimeOptionsManager() runtime.OptionsManager {
+	initRuntimeOpts := runtime.NewOptions().
+		SetWriteValuesPerMetricLimitPerSecond(c.WriteValuesPerMetricLimitPerSecond).
+		SetWriteNewMetricNoLimitWarmupDuration(c.WriteNewMetricNoLimitWarmupDuration)
+	return runtime.NewOptionsManager(initRuntimeOpts)
+}
+
+// WatchRuntimeOptionChanges watches runtime option updates.
+func (c RuntimeOptionsConfiguration) WatchRuntimeOptionChanges(
+	client client.Client,
+	runtimeOptsManager runtime.OptionsManager,
+	placementManager aggregator.PlacementManager,
+	logger log.Logger,
+) {
+	kvOpts, err := c.KVConfig.NewOptions()
+	if err != nil {
+		logger.Errorf("unable to create kv config options: %v", err)
+		return
+	}
+	store, err := client.Store(kvOpts)
+	if err != nil {
+		logger.Errorf("unable to create kv store: %v", err)
+		return
+	}
+
+	var (
+		valueLimitKey                = c.WriteValuesPerMetricLimitPerSecondKey
+		defaultValueLimit            = c.WriteValuesPerMetricLimitPerSecond
+		valueLimit                   int64
+		valueLimitCh                 <-chan struct{}
+		newMetricClusterLimitKey     = c.WriteNewMetricLimitClusterPerSecondKey
+		defaultNewMetricClusterLimit = c.WriteNewMetricLimitClusterPerSecond
+		newMetricClusterLimit        int64
+		newMetricPerShardLimit       int64
+		newMetricLimitCh             <-chan struct{}
+	)
+	valueLimit, err = retrieveLimit(valueLimitKey, store, defaultValueLimit)
+	if err != nil {
+		logger.Errorf("unable to retrieve per-metric write value limit from kv: %v", err)
+	}
+	logger.Infof("current write value limit per second is: %d", valueLimit)
+
+	newMetricClusterLimit, err = retrieveLimit(newMetricClusterLimitKey, store, defaultNewMetricClusterLimit)
+	if err == nil {
+		newMetricPerShardLimit, err = clusterLimitToPerShardLimit(newMetricClusterLimit, placementManager)
+	}
+	if err != nil {
+		logger.Errorf("unable to determine per-shard write new metric limit: %v", err)
+	}
+	logger.Infof("current write new metric limit per shard per second is: %d", newMetricPerShardLimit)
+
+	runtimeOpts := runtime.NewOptions().
+		SetWriteNewMetricNoLimitWarmupDuration(c.WriteNewMetricNoLimitWarmupDuration).
+		SetWriteValuesPerMetricLimitPerSecond(valueLimit).
+		SetWriteNewMetricLimitPerShardPerSecond(newMetricPerShardLimit)
+	runtimeOptsManager.SetRuntimeOptions(runtimeOpts)
+
+	valueLimitWatch, err := store.Watch(valueLimitKey)
+	if err != nil {
+		logger.Errorf("unable to watch per-metric write value limit: %v", err)
+	} else {
+		valueLimitCh = valueLimitWatch.C()
+	}
+	newMetricLimitWatch, err := store.Watch(newMetricClusterLimitKey)
+	if err != nil {
+		logger.Errorf("unable to watch cluster-wide write new metric limit: %v", err)
+	} else {
+		newMetricLimitCh = newMetricLimitWatch.C()
+	}
+	// If watch creation failed for both, we return immediately.
+	if valueLimitCh == nil && newMetricLimitCh == nil {
```

Comment:
should this check be ||?

Reply:
No, if we can't watch one, at least we can watch the other.